credentials/generate-revocation-set.py

```diff
@@ -0,0 +1,243 @@
+# Usage:
+#     python ./credentials/generate-revocation-set.py --help
+
+import base64
+import json
+import subprocess
+import sys
+from enum import Enum
+
+import click
+import requests
+from click_option_group import RequiredMutuallyExclusiveOptionGroup, optgroup
+from cryptography import x509
+
+
+class RevocationType(Enum):
+    CRL = 1
+
+
+OID_VENDOR_ID = x509.ObjectIdentifier("1.3.6.1.4.1.37244.2.1")
+OID_PRODUCT_ID = x509.ObjectIdentifier("1.3.6.1.4.1.37244.2.2")
+
+PRODUCTION_NODE_URL = "https://on.dcl.csa-iot.org:26657"
+PRODUCTION_NODE_URL_REST = "https://on.dcl.csa-iot.org"
+TEST_NODE_URL_REST = "https://on.test-net.dcl.csa-iot.org"
+
+
+def use_dcld(dcld, production, cmdlist):
+    return [dcld] + cmdlist + (['--node', PRODUCTION_NODE_URL] if production else [])
+
+
+def extract_single_integer_attribute(subject, oid):
+    attribute_list = subject.get_attributes_for_oid(oid)
+
+    if len(attribute_list) == 1:
+        if attribute_list[0].value.isdigit():
+            return int(attribute_list[0].value)
+
+    return None
+
+
+@click.command()
+@click.help_option('-h', '--help')
+@optgroup.group('Input data sources', cls=RequiredMutuallyExclusiveOptionGroup)
+@optgroup.option('--use-main-net-dcld', type=str, default='', metavar='PATH', help="Location of `dcld` binary, to use `dcld` for mirroring MainNet.")
+@optgroup.option('--use-test-net-dcld', type=str, default='', metavar='PATH', help="Location of `dcld` binary, to use `dcld` for mirroring TestNet.")
+@optgroup.option('--use-main-net-http', is_flag=True, type=str, help="Use RESTful API with HTTPS against public MainNet observer.")
+@optgroup.option('--use-test-net-http', is_flag=True, type=str, help="Use RESTful API with HTTPS against public TestNet observer.")
+@optgroup.group('Optional arguments')
+@optgroup.option('--output', default='sample_revocation_set_list.json', type=str, metavar='FILEPATH', help="Output filename (default: sample_revocation_set_list.json)")
+def main(use_main_net_dcld, use_test_net_dcld, use_main_net_http, use_test_net_http, output):
+    """DCL PAA mirroring tools"""
+
+    production = False
+    dcld = use_test_net_dcld
+
+    if len(use_main_net_dcld) > 0:
+        dcld = use_main_net_dcld
+        production = True
+
+    use_rest = use_main_net_http or use_test_net_http
+    if use_main_net_http:
+        production = True
+
+    rest_node_url = PRODUCTION_NODE_URL_REST if production else TEST_NODE_URL_REST
+
+    if use_rest:
+        revocation_point_list = requests.get(f"{rest_node_url}/dcl/pki/revocation-points").json()["PkiRevocationDistributionPoint"]
+    else:
+        cmdlist = ['config', 'output', 'json']
+        subprocess.Popen([dcld] + cmdlist)
+
+        cmdlist = ['query', 'pki', 'all-revocation-points']
+
+        cmdpipe = subprocess.Popen(use_dcld(dcld, production, cmdlist), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        revocation_point_list = json.loads(cmdpipe.stdout.read())["PkiRevocationDistributionPoint"]
```

Comment:
Extract this to a helper function. Can be done as a follow-up.

Reply:
Added comments in the src file to capture the follow-up work.